### PR TITLE
txn: resolve left secondary lock to avoid dead loop

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What problem does this PR solve? 

Resolving possible left secondary pessimistic lock to avoid dead loop

### What is changed and how it works?

1. Add pessimistic lock  resolve logic for lock resolver
2. Change resolve logic in `mvcc_leveldb`
3. Add check conflict logic in the pessimistic lock processing 
4. **Change `pessimisticLockMaxBackoff` and `pessimisticRollbackMaxBackoff` to 20s, and make it
the same as `managedLockTTL`**

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

 - Need to cherry-pick to the release branch

Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 1]
```diff
@@ -1129,10 +1129,18 @@ func (mvcc *MVCCLevelDB) CheckTxnStatus(primaryKey []byte, lockTS, callerStartTS
 
 	if rollbackIfNotExist {
 		batch := &leveldb.Batch{}
-		if err1 := rollbackLock(batch, primaryKey, lockTS); err1 != nil {
-			err = errors.Trace(err1)
+		tomb := mvccValue{
```
Comment (coocood) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: Why not use `rollbackLock` directly?

Comment (coocood) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: We can extract a common method to reduce duplicated code.

**Comment by cfzjywxk** [comment]
The unit test running together, sometimes `TestSingleStatementRollback` will fail like this
```
[2020-02-14T05:02:09.823Z] ----------------------------------------------------------------------
[2020-02-14T05:02:09.823Z] FAIL: pessimistic_test.go:179: testPessimisticSuite.TestSingleStatementRollback
[2020-02-14T05:02:09.823Z] 
[2020-02-14T05:02:09.823Z] pessimistic_test.go:210:
[2020-02-14T05:02:09.823Z]     tk.MustExec("update single_statement set v = v + 1")
[2020-02-14T05:02:09.823Z] /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:183:
[2020-02-14T05:02:09.823Z]     tk.c.Assert(err, check.IsNil, check.Commentf("sql:%s, %v, error stack %v", sql, args, errors.ErrorStack(err)))
[2020-02-14T05:02:09.823Z] ... value *errors.withStack = [tikv:1205]Lock wait timeout exceeded; try restarting transaction ("[tikv:1205]Lock wait timeout exceeded; try restarting transaction")
[2020-02-14T05:02:09.823Z] ... sql:update single_statement set v = v + 1, [], error stack [tikv:1205]Lock wait timeout exceeded; try restarting transaction
[2020-02-14T05:02:09.823Z] github.com/pingcap/errors.AddStack
[2020-02-14T05:02:09.823Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20190809092503-95897b64e011/errors.go:174
[2020-02-14T05:02:09.823Z] github.com/pingcap/errors.Trace
[2020-02-14T05:02:09.823Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20190809092503-95897b64e011/juju_adaptor.go:15
[2020-02-14T05:02:09.823Z] github.com/pingcap/tidb/store/tikv.actionPessimisticLock.handleSingleBatch
[2020-02-14T05:02:09.823Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:769
[2020-02-14T05:02:09.823Z] github.com/pingcap/tidb/store/tikv.(*batchExecutor).startWorker.func1
[2020-02-14T05:02:09.823Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:1285
[2020-02-14T05:02:09.823Z] runtime.goexit
[2020-02-14T05:02:09.823Z] 	/usr/local/go/src/runtime/asm_amd64.s:1357
```
Running seperately it will always succeed...

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 2]
```diff
@@ -1128,8 +1136,9 @@ func (mvcc *MVCCLevelDB) CheckTxnStatus(primaryKey []byte, lockTS, callerStartTS
 	// written before the primary lock.
 
 	if rollbackIfNotExist {
+		// Write rollback record, but not delete the lock
```
Comment (tiancaiamao) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: Why not delete the lock, to keep consistent with TiKV?

File: store/tikv/lock_resolver.go [hunk 3]
```diff
@@ -410,10 +410,6 @@ func (lr *LockResolver) getTxnStatusFromLock(bo *Backoffer, l *Lock, callerStart
 			return TxnStatus{}, err
 		}
 
-		if l.LockType == kvrpcpb.Op_PessimisticLock {
```
Comment (tiancaiamao) [inline store/tikv/lock_resolver.go]: I guess TestOptimisticConflicts is not stable after this change?

File: store/mockstore/mocktikv/mvcc_leveldb.go (hunk 2 above)
Comment (cfzjywxk) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: There may exist one lock on the primary key, and this lock belongs to other transactions, like the transaction which raises the resolve action itself, we've already checked the lock for the primary  key and `lockTS`

Comment (lysu) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: could we add this explanation as code comments, other code readers may have had the same question in mind

File: store/tikv/lock_resolver.go [hunk 4]
```diff
@@ -308,7 +308,12 @@ func (lr *LockResolver) ResolveLocks(bo *Backoffer, callerStartTS uint64, locks
 				cleanTxns[l.TxnID] = cleanRegions
 			}
 
-			err = lr.resolveLock(bo, l, status, cleanRegions)
+			if l.LockType == kvrpcpb.Op_PessimisticLock {
+				//logutil.Logger(bo.ctx).Warn("found expired pessimistic lock", zap.Stringer("lock", l))
```
Comment (jackysp) [inline store/tikv/lock_resolver.go]: please remove this debug comment

File: store/mockstore/mocktikv/mvcc_leveldb.go (hunk 2 above)
Comment (tiancaiamao) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: > There may exist one lock on the primary key, and this lock belongs to other transactions

And we're rolling back it.

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 5]
```diff
@@ -645,7 +648,7 @@ func (mvcc *MVCCLevelDB) Prewrite(req *kvrpcpb.PrewriteRequest) []error {
 	return errs
 }
 
-func checkConflictValue(iter *Iterator, m *kvrpcpb.Mutation, startTS uint64) error {
+func checkConflictValue(iter *Iterator, m *kvrpcpb.Mutation, startTS uint64, txnTS uint64) error {
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: Or maybe `callerTS` is better, for optimistic txn `startTS` is used, for pessimistic txn `forUpdateTS` is used.

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 6]
```diff
@@ -1128,8 +1170,12 @@ func (mvcc *MVCCLevelDB) CheckTxnStatus(primaryKey []byte, lockTS, callerStartTS
 	// written before the primary lock.
 
 	if rollbackIfNotExist {
+		// Write rollback record, but not delete the lock on the primary key. There may exist lock which has
```
Comment (tiancaiamao) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: HINT: This function works on the primary key!

File: store/mockstore/mocktikv/mvcc_leveldb.go (hunk 2 above)
Comment (tiancaiamao) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: Finally I understand here.

File: store/tikv/lock_resolver.go [hunk 7]
```diff
@@ -410,22 +414,21 @@ func (lr *LockResolver) getTxnStatusFromLock(bo *Backoffer, l *Lock, callerStart
...
 
-		if l.LockType == kvrpcpb.Op_PessimisticLock {
-			return TxnStatus{ttl: l.TTL}, nil
-		}
-
 		// Handle txnNotFound error.
 		// getTxnStatus() returns it when the secondary locks exist while the primary lock doesn't.
 		// This is likely to happen in the concurrently prewrite when secondary regions
 		// success before the primary region.
 		if err := bo.Backoff(boTxnNotFound, err); err != nil {
```
Comment (coocood) [inline store/tikv/lock_resolver.go]: The `ManagedLockTTL` is 20s, and `pessimisticLockMaxBackoff` is 10s，the transaction will exceed the max backoff and return error before the TTL reached.

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 8]
```diff
@@ -655,9 +659,9 @@ func checkConflictValue(iter *Iterator, m *kvrpcpb.Mutation, startTS uint64) err
 	}
 
 	// Note that it's a write conflict here, even if the value is a rollback one.
-	if dec.value.commitTS >= startTS {
+	if dec.value.commitTS >= callerTS {
```
Comment (coocood) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: Should be `startTS`?

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 9]
```diff
@@ -645,7 +648,8 @@ func (mvcc *MVCCLevelDB) Prewrite(req *kvrpcpb.PrewriteRequest) []error {
 	return errs
 }
 
-func checkConflictValue(iter *Iterator, m *kvrpcpb.Mutation, startTS uint64) error {
+// callerTS is startTS for optimistic transactions, and forUpdateTS for pessimistic transactions
+func checkConflictValue(iter *Iterator, m *kvrpcpb.Mutation, callerTS uint64, startTS uint64) error {
```
Comment (coocood) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: The name `callerTS` may be confused with `callerStartTS` used in `CheckTxnStatus`.

Comment (cfzjywxk) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: Or just use `forUpdateTS` ?

Comment (coocood) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: OK

**Comment by coocood** [comment]
LGTM

File: store/tikv/backoff.go [hunk 10]
```diff
@@ -218,8 +218,8 @@ const (
 	scatterRegionBackoff           = 20000
 	waitScatterRegionFinishBackoff = 120000
 	locateRegionMaxBackoff         = 20000
-	pessimisticLockMaxBackoff      = 10000
```
Comment (lysu) [inline store/tikv/backoff.go]: also need pay attention and test to take look `WithVars` https://github.com/pingcap/tidb/blob/d20195a3a5205bce876afa02f0954d682476918f/store/tikv/txn.go#L385
will make real max be `pessimisticLockMaxBackoff x 2` or maybe in this situation we should ignore `BackOffWeight`? @jackysp

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 11]
```diff
@@ -683,6 +686,28 @@ func checkConflictValue(iter *Iterator, m *kvrpcpb.Mutation, startTS uint64) err
...
+		// Check if rollback write record(startTS, startTS, typeRollback) exists on key
+		info, ok, err1 := getTxnCommitInfo(iter, m.Key, startTS)
+		if err1 != nil {
+			err = errors.Trace(err1)
+			return err
+		}
+		if ok && info.commitTS == startTS {
+			logutil.BgLogger().Warn("rollback value found",
+				zap.Uint64("txnID", startTS),
+				zap.Int32("info.valueType", int32(info.valueType)),
```
Comment (lysu) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: it seems this value always be `typeRollback`, maybe better change `info.*` to `rollbacked.*`~?

and add type info to above panic message `panic(fmt.Sprintf("invalid value type %v not rollback", info.valueType))`

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 12]
```diff
@@ -1128,8 +1161,12 @@ func (mvcc *MVCCLevelDB) CheckTxnStatus(primaryKey []byte, lockTS, callerStartTS
 	// written before the primary lock.
 
 	if rollbackIfNotExist {
+		// Write rollback record, but not delete the lock on the primary key. There may exist lock which has
+		// different lock.startTS with input lockTS, for example the primary key could be already
+		// locked by the caller transaction, deleting this key will mistakenly delete the lock on
+		// primary key, see case TestSingleStatementRollbackSecondary in session_test suite for example
```
Comment (lysu) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: this seems need update comments after rename `TestSingleStatementRollback`, and do we need change this after choose method2 that no longer CheckTxnStatus(rollbackIfNotExist=true)?

File: store/tikv/backoff.go (hunk 10 above)
Comment (cfzjywxk) [reply store/tikv/backoff.go]: Transaction related `NewBackoffer` call are mostly  used with `WithVars` and will be affected by `BackOffWeight`,  this change will make pessimistic max backoff time doubled as before. Are there any corner cases using `BackOffWeight` ? @jackysp

**Comment by tiancaiamao** [comment]
LGTM

**Comment by cfzjywxk** [comment]
@jackysp  @lysu PTAL

File: store/tikv/backoff.go (hunk 10 above)
Comment (coocood) [reply store/tikv/backoff.go]: 10s is too short compares to other maxBackoff, I think it's doesn't hurt to increase to 20s

Comment (jackysp) [reply store/tikv/backoff.go]: We could discuss it this afternoon.

**Comment by coocood** [comment]
@lysu PTAL

File: store/tikv/lock_resolver.go [hunk 13]
```diff
@@ -573,3 +576,47 @@ func (lr *LockResolver) resolveLock(bo *Backoffer, l *Lock, status TxnStatus, cl
...
 }
+
+func (lr *LockResolver) resolvePessimisticLock(bo *Backoffer, l *Lock, cleanRegions map[RegionVerID]struct{}) error {
+	tikvLockResolverCountWithResolveLocks.Inc()
+	for {
+		loc, err := lr.store.GetRegionCache().LocateKey(bo, l.Key)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if _, ok := cleanRegions[loc.Region]; ok {
```
Comment (lysu) [inline store/tikv/lock_resolver.go]: maybe we should remove this in next pr.

it seems txn first resolve a optimistic lock then can not resolve pessimistic lock, if mix optimistic/pessmistic

**Pull request merged**